### PR TITLE
Fixing properties

### PR DESCRIPTION
**Pull request by William-TogetherScience**

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -46,4 +46,4 @@ jobs:
 
       - name: Run tests
         shell: bash
-        run: npm run test
```
Comment (William-TogetherScience) [inline .github/workflows/main.yml:49]: Add line break

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "command": "npm run quick-test",
+            "name": "quick tests",
+            "request": "launch",
+            "type": "node-terminal"
+        }
+    ]
+}
```
Comment (William-TogetherScience) [inline .vscode/launch.json]: Add line break

File: branch.txt
Comment (William-TogetherScience) [inline branch.txt:1]: Add to gitignore

File: ts-port/core/assumptions.ts [hunk 3]
```diff
@@ -161,12 +167,7 @@ function _ask(fact: any, obj: any) {
         let fact_i_value = undefined;
         let handler_i = handler_map.get(fact_i);
         if (typeof handler_i !== "undefined") {
-            handler_i = handler_i.name;
-            if (obj[handler_i]) {
```
Comment (William-TogetherScience) [inline ts-port/core/assumptions.ts:165]: Double check (with sympy src)

File: ts-port/core/assumptions.ts [hunk 4]
```diff
@@ -224,39 +224,41 @@ class ManagedProperties {
             }
         }
 
-        // Keep track of the explicit assumptions for all registered classes.
-        // For a given class, this looks like the assumptions for all of its
-        // superclasses since we register classes top-down.
-        this.all_explicit_assumptions.merge(local_defs);
+        const all_defs = new HashDict()
+        for (const base of Util.getSupers(cls).reverse()) {
```
Comment (William-TogetherScience) [inline ts-port/core/assumptions.ts]: go through with counter (avoid reverse())

File: ts-port/core/assumptions.ts [hunk 5]
```diff
@@ -224,39 +224,41 @@ class ManagedProperties {
...
-        cls._explicit_class_assumptions = this.all_explicit_assumptions;
-        cls.default_assumptions = new StdFactKB(this.all_explicit_assumptions);
+        // Set class properties for assume_defined
+        cls._explicit_class_assumptions = all_defs
+        cls.default_assumptions = new StdFactKB(all_defs);
 
         // Add default assumptions as class properties
         for (const item of cls.default_assumptions.entries()) {
-            cls[as_property(item[0])] = item[1];
+            if (item[0].includes("is")) {
```
Comment (William-TogetherScience) [inline ts-port/core/assumptions.ts]: Is this necessary?

File: ts-port/core/basic.ts [hunk 6]
```diff
@@ -295,7 +306,7 @@ const _Basic = (superclass: any) => class _Basic extends superclass {
                 let rv;
                 if (cls.constructor.name === "Mul" || cls.constructor.name === "Add") {
                     rv = new cls.constructor(true, true, ...args);
-                } else if (cls.constructor.name === "Pow") {
```
Comment (William-TogetherScience) [inline ts-port/core/basic.ts]: Why?

File: ts-port/core/facts.ts [hunk 7]
```diff
@@ -74,6 +72,18 @@ function _as_pair(atom: any) {
     }
 }
 
+
+function _as_pairv2(atom: any) {
```
Comment (William-TogetherScience) [inline ts-port/core/facts.ts:76]: Look at booleans (and Logic.True and Logic.False)

File: ts-port/core/facts.ts [hunk 8]
```diff
@@ -394,7 +410,8 @@ class Prover {
 
             for (let bidx = 0; bidx < b.args.length; bidx++) {
                 const barg = b.args[bidx];
-                const brest = [...b.args].splice(bidx, 1);
```
Comment (William-TogetherScience) [inline ts-port/core/facts.ts:397]: brest does not return

File: ts-port/core/facts.ts [hunk 9]
```diff
@@ -599,37 +618,38 @@ class FactKB extends HashDict {
 
             // --- alpha chains ---
             for (const item of facts) {
-                const k = item[0];
-                const v = item[1];
+                let k, v;
+                if (item instanceof Implication) {
+                    k = item.p;
+                    v = item.q
```
Comment (William-TogetherScience) [inline ts-port/core/facts.ts]: Make consistent (always implication)
Throw error if its an array?

File: ts-port/core/logic.ts [hunk 10]
```diff
@@ -309,20 +310,20 @@ class Logic {
...
-            return Not.__new__(Not.prototype, arg);
+            return Not.New(arg);
         },
     };
 
     args: any[];
 
     constructor(...args: any[]) {
-        this.args = args;
+        this.args = [...args].flat()
```
Comment (William-TogetherScience) [inline ts-port/core/logic.ts]: Is this necessary? Avoid making unncessary copies

File: ts-port/core/logic.ts [hunk 11]
```diff
@@ -532,18 +536,15 @@ class AndOr_Base extends Logic {
             }
             res.push(arg);
         }
-        return res;
+        return res.flat();
```
Comment (William-TogetherScience) [inline ts-port/core/logic.ts]: See if this makes a copy

File: ts-port/core/mul.ts [hunk 12]
```diff
@@ -471,11 +474,11 @@ export class Mul extends mix(base).with(Expr, AssocOp) {
...
                 const new_c_part = [];
                 for (const t of c_part) {
-                    if (t.is_extended_negative) {
+                    if (t.is_extended_positive()) {
                         continue;
                     }
-                    if (t.is_extended_negative) {
-                        coeff_sign *= -1;
+                    if (t.is_extended_negative()) {
+                        coeff_sign = -1;
```
Comment (William-TogetherScience) [inline ts-port/core/mul.ts]: Typo??

File: ts-port/core/numbers.ts [hunk 13]
```diff
@@ -785,7 +793,7 @@ class Integer extends Rational {
         const [x, xexact] = int_nthroot(Math.abs(this.p), expt.q);
         if (xexact) {
             let result = new Integer((x as number)**Math.abs(expt.p));
-            if (this.is_negative()) {
+            if (this.is_negative() == true) {
```
Comment (William-TogetherScience) [inline ts-port/core/numbers.ts]: Not necessary

File: ts-port/core/utility.ts [hunk 14]
```diff
@@ -195,6 +199,22 @@ class Util {
...
     }
+
+    static splitLogicStr(str: string): any[] {
+        const sep = " ";
+        const max_splits = 3;
+        
+        const orig_split = str.split(" ", 10)
+        if (orig_split.length == 3) {
+            return orig_split;
+        } else {
```
Comment (William-TogetherScience) [inline ts-port/core/utility.ts]: Use regex code (in teams)

**Review by gunnarmein-ts — APPROVED** [review]

**Pull request merged**